lib/zendesk_apps_support/validations/manifest.rb

```diff
@@ -39,25 +39,14 @@ def collate_manifest_errors(package)
           errors << oauth_error(manifest)
           errors << default_locale_error(manifest, package)
 
-          if manifest.marketing_only?
-            errors.push(*marketing_only_errors(manifest))
-          else
-            errors << parameters_error(manifest)
-            errors << invalid_hidden_parameter_error(manifest)
-            errors << invalid_type_error(manifest)
-            errors << too_many_oauth_parameters(manifest)
-            errors << name_as_parameter_name_error(manifest)
-          end
+          errors << validate_parameters(manifest)
 
           if manifest.requirements_only? || manifest.marketing_only?
             errors << ban_location(manifest)
             errors << ban_framework_version(manifest)
           else
-            errors << missing_location_error(package)
-            errors << invalid_location_error(package)
-            errors << invalid_v1_location(package)
+            errors << validate_location(package)
```

Comment:
Because `validate_location` is returning an array I'm surprised that you don't need to do `.concat` here... I see down below that we're doing `flatten.compact` on the whole thing.. That actually gives us a lot of room to improve this whole thing. We could remove every instance of `<<` here by instead making the whole method an array literal with method calls inside it. Like:

``` ruby
[
  oauth_error(manifest),
  default_locale_error(manifest, package),
  validate_parameters(manifest),
  if manifest.requirements_only? || manifest.marketing_only?
    [
      ban_location(manifest),
      ban_framework_version(manifest)
    ]
  else
    [ ... ]
  end,
  ...
]
```

Because `if` is an expression in ruby it should work nicely, and remove a lot of this duplication. I should've seen this earlier, but I had assumed there was some structure to the errors that needed to be maintained.

Up to you if you'd like to try this style of change.

Reply:
We did this together and use the same approach to refactor other methods